src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -7925,43 +7955,54 @@ private BoundExpression BindIndexedPropertyAccess(SyntaxNode syntax, BoundExpres
                             // note: implicit copy check on the indexer accessor happens in CheckPropertyValueKind
-                            patternIndexerAccess = new BoundIndexOrRangePatternIndexerAccess(
-                                syntax,
-                                receiverOpt,
-                                lengthOrCountProperty,
-                                property,
-                                BindToNaturalType(argument, diagnostics),
-                                property.Type);
-                            break;
+                            patternSymbol = property;
+                            returnType = property.Type;
+                            GetWellKnownTypeMember(WellKnownMember.System_Index__GetOffset, diagnostics, syntax: syntax);
+                            return true;
                         }
                     }
                 }
             }
-            else if (receiverType.SpecialType == SpecialType.System_String)
+            else if (TryFindImplicitRangeIndexerPattern(syntax, receiverOpt, receiverType,
+                    out MethodSymbol? sliceMethod, diagnostics, lookupResult, ref discardedUseSiteInfo))
+            {
+                patternSymbol = sliceMethod;
+                returnType = sliceMethod.ReturnType;
+                CheckImplicitThisCopyInReadOnlyMember(receiverOpt, lengthOrCountProperty.GetMethod, diagnostics);
+                return true;
+            }
+
+            patternSymbol = null;
+            returnType = null;
+            return false;
+        }
+
+        private bool TryFindImplicitRangeIndexerPattern(
+            SyntaxNode syntax, BoundExpression? receiverOpt,
+            TypeSymbol receiverType,
+            [NotNullWhen(true)] out MethodSymbol? patternSymbol,
+            BindingDiagnosticBag diagnostics, LookupResult lookupResult,
+            ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            patternSymbol = null;
+
+            if (receiverType.SpecialType == SpecialType.System_String)
             {
-                Debug.Assert(argIsRange);
                 // Look for Substring
                 var substring = (MethodSymbol)Compilation.GetSpecialTypeMember(SpecialMember.System_String__Substring);
                 if (substring is object)
                 {
-                    patternIndexerAccess = new BoundIndexOrRangePatternIndexerAccess(
-                        syntax,
-                        receiverOpt,
-                        lengthOrCountProperty,
-                        substring,
-                        BindToNaturalType(argument, diagnostics),
-                        substring.ReturnType);
-                    checkWellKnown(WellKnownMember.System_Range__get_Start);
-                    checkWellKnown(WellKnownMember.System_Range__get_End);
+                    patternSymbol = substring;
+                    checkWellKnownMembers();
```

Comment:
There's a behavior change here. We're now checking 3 well-known members instead of 2 previously. Is this intentional?
Ditto for other usages of `checkWellKnownMembers`.

Reply:
>I don't see the problem with lowering having some of its own additional checks for well-known types when they are needed.

This is just a move. I'd be reluctant to change logic in this method because it will affect range indexer binding. Can be done in a separate PR though.

Reply:
I didn't follow. It seems more than just a move.
Previously we were just checking for Range.Start and Range.End here, but now we're additionally checking for Index.Offset, no?

Reply:
It's checked at the very end of the original method which is always reachable for success case.

Reply:
Ah, I'd missed that. Thanks